<a href="https://colab.research.google.com/github/joiepark/assignment_esaa/blob/main/YB%205.1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 사이킷런 소개와 특징



*   사이킷런(scikit-learn)은 파이썬 머신러닝 라이브러리 중 가장 많이 사용되는 라이브러리이다.
*   사이킷런은 파이썬 기반의 다른 머신러닝 패키지도 사이킷런 스타일의 API를 지향할 정도로 쉽고 가장 파이썬스러운 API를 제공한다. 또, 머신러닝을 위한 매우 다양한 알고리즘과 개발을 위한 편리한 프레임워크와 API를 제공한다. 오랜 시간 실전 환경에서 검증되었으며, 매우 많은 환경에서 사용되는 성숙한 라이브러리다.




In [2]:
pip install scikit-learn==1.6.1

In [3]:
import sklearn

print(sklearn.__version__)

1.6.1


## 첫 번째 머신러닝 만들어 보기 - 붓꽃 품종 예측하기

붓꽃 데이터 세트로 붓꽃의 품종을 분류하는 것 : 붓꽃 데이터 세트는 꽃잎의 길이와 너비, 꽃받침의 길이와 너비 피처를 기반으로 꽃의 품종을 예측하기 위한 것이다.



*   분류(Classification)는 대표적인 **지도학습(Supervised Learning)** 방법의 하나이다. 지도학습은 학습을 위한 다양한 피처와 분류 결정값인 레이블(Label) 데이터로 모델을 학습한 뒤, 별도의 테스트 데이터 세트에서 미지의 레이블을 예측한다. 즉, 지도학습은 명확한 정답이 주어진 데이터를 먼저 학습한 뒤 미지의 정답을 예측하는 방식이다. 이때 학습을 위해 주어진 데이터 세트를 **학습 데이터 세트**, 머신러닝 모델의 예측 성능을 평가하기 위해 별도로 주어진 데이터 세트를 **테스트 데이터 세트**로 지칭한다.


*   사이킷런 패키지 내의 모듈명은 sklearn으로 시작하는 명명규칙이 있다. **sklearn.datasets** 내의 모듈은 사이킷런에서 자체적으로 제공하는 데이터 세트를 생성하는 모듈의 모임이다. **sklearn.tree** 내의 모듈은 트리 기반 ML 알고리즘을 구현한 클래스의 모임이다. **sklearn.model_selection**은 학습 데이터와 검증 데이터, 예측 데이터로 데이터를 분리하거나 최적의 하이퍼 파라미터로 평가하기 위한 다양한 모듈의 모임이다.
*   **하이퍼 파라미터**란 머신러닝 알고리즘별로 최적의 학습을 위해 직접 입력하는 파라미터들을 통칭한다. 하이퍼 파라미터를 통해 머신러닝 알고리즘의 성능을 튜닝할 수 있다.



In [4]:
from sklearn.datasets import load_iris   ## 붓꽃 데이터 세트를 생성
from sklearn.tree import DecisionTreeClassifier   ## ML 알고리즘은 의사 결정 트리(Decision Tree) 알고리즘
from sklearn.model_selection import train_test_split   ## 데이터 세트를 학습 데이터와 테스트 데이터로 분리

In [5]:
import pandas as pd

# 붓꽃 데이터 세트를 로딩한다
iris = load_iris()

# iris.data는 Iris 데이터 세트에서 피처(feature)만으로 된 데이터를 numpy로 가지고 있다
iris_data = iris.data

# iris.target은 붓꽃 데이터 세트에서 레이블(결정 값) 데이터를 numpy로 가지고 있다
iris_label = iris.target
print('iris target값:', iris_label)
print('iris target명:', iris.target_names)

# 붓꽃 데이터 세트를 자세히 보기 위해 DataFrame으로 변환한다
iris_df = pd.DataFrame(data=iris_data, columns=iris.feature_names)
iris_df['label'] = iris.target
iris_df.head(3)

iris target값: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
iris target명: ['setosa' 'versicolor' 'virginica']


,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),label
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0


피처에는 sepal length, sepal width, petal length, petal width가 있다. 레이블(Label, 결정값)은 0, 1, 2 세 가지 값으로 되어 있으며 0이 Setosa 품종, 1이 versicolor 품종, 2가 virginica 품종을 의미한다.


*   학습용 데이터와 테스트용 데이터는 반드시 분리해야 한다. 학습 데이터로 학습된 모델이 얼마나 뛰어난 성능을 가지는지 평가하려면 테스트 데이터 세트가 필요하기 때문이다. 이를 위해 사이킷런은 **train_test_split()** API를 제공한다. train_test_split()을 이용하면 학습 데이터와 테스트 데이터를 test_size 파라미터 입력값의 비율로 쉽게 분할한다. (e.g. test_size=0.2로 입력 파라미터를 설정하면 전체 데이터 중 테스트 데이터가 20%, 학습 데이터가 80%로 데이터를 분할한다.)
*   train_test_split()의 첫 번째 파라미터는 **피처 데이터 세트**이고, 두 번째 파라미터는 **레이블(Label) 데이터 세트**이다. 그리고 **test_size**는 전체 데이터 세트 중 테스트 테이트 세트의 비율이다. 마지막으로 **random_state**는 호출할 때마다 같은 학습/테스트용 데이터 세트를 생성하기 위해 주어지는 난수 발생 값이다. random_state는 random값을 만드는 seed와 같은 의미이므로 숫자 자체는 어떤 값을 지정해도 상관없다.
*   학습 데이터를 확보한 이후에는 이 데이터를 기반으로 머신러닝 분류 알고리즘의 하나인 **의사 결정 트리**를 이용해 학습과 예측을 수행할 수 있다. 먼저 사이킷런의 의사 결정 트리 클래스인 **DecisionTreeClassifier**를 객체로 생성하고, 생성된 객체의 **fit()** 메서드에 학습용 피처 데이터 속성과 결정값 데이터 세트를 입력해 호출하면 학습을 수행한다.





In [6]:
X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label,
                                                    test_size=0.2, random_state=11)

In [7]:
# DecisionTreeClassifier 객체를 생성한다
dt_clf = DecisionTreeClassifier(random_state=11)   ## DecisionTreeClassifier 객체 생성 시 입력된 random_state 역시 예제 코드를 수행할 때마다 동일한 학습/예측 결과를 출력하기 위한 용도로만 사용된다

# 학습을 수행한다
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier(random_state=11)



*   DecisionTreeClassifier 객체가 학습이 완료되면 이 객체를 이용해 예측을 수행한다. 예측은 반드시 학습 데이터가 아닌 다른 데이터를 이용해야 하며, 일반적으로 테스트 데이터 세트를 이용한다. DecisionTreeClassifier 객체의 **predict()** 메서드에 테스트용 피처 데이터 세트를 입력해 호출하면 학습된 모델 기반에서 테스트 데이터 세트에 대한 예측값을 반환하게 된다.
*   예측 결과를 기반으로 DecisionTreeClassifier의 예측 성능을 평가할 수 있다. 일반적으로 머신러닝 모델의 성능 평가 방법은 여러 가지가 있다. 그 중 **정확도**는 예측 결과가 실제 레이블 값과 얼마나 정확하게 맞는지를 평가하는 지표이다. 사이킷런은 정확도 측정을 위해 **accuracy_score()** 함수를 제공한다. accuracy_score()의 첫 번째 파라미터로 실제 레이블 데이터 세트, 두 번째 파라미터로 예측 레이블 데이터 세트를 입력하면 된다.



In [8]:
# 학습이 완료된 DecisionTreeClassifier 객체에서 테스트 데이터 세트로 예측을 수행한다
pred = dt_clf.predict(X_test)

In [9]:
from sklearn.metrics import accuracy_score
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))

예측 정확도: 0.9333


## 사이킷런의 기반 프레임워크 익히기

### Estimator 이해 및 fit(), predict() 메서드



*   사이킷런은 ML 모델 학습을 위해 fit()을, 학습된 모델의 예측을 위해 predict() 메서드를 제공한다. 지도학습의 주요 두 축인 분류(Classification)와 회귀(Regression)의 다양한 알고리즘을 구현한 모든 사이킷런 클래스는 fit()과 predict()만을 이용해 간단하게 학습과 예측 결과를 반환한다. 사이킷런에서는 분류 알고리즘을 구현한 클래스를 **Classifier**로, 회귀 알고리즘을 구현한 클래스를 **Regressor**로 지칭한다. 이들 Classifier와 Regressor를 합쳐서 **Estimator 클래스**라고 부른다. 즉, 지도학습의 모든 알고리즘을 구현한 클래스를 통칭해서 Estimator라고 부른다. evaluation 함수나 하이퍼 파라미터 튜닝을 지원하는 클래스의 경우 이 Estimator을 인자로 받는다. 인자로 받은 Estimator에 대해 함수 내에서 이 Estimator의 fit()과 predict()를 호출해 평가를 하거나 하이퍼 파라미터 튜닝을 수행하는 것이다.
*   사이킷런에서 비지도학습인 차원 축소, 클러스터링, 피처 추출(Feature Extraction) 등을 구현한 클래스 역시 대부분 fit()과 transform()을 적용한다. **비지도학습과 피처 추출에서 fit()**은 지도학습의 fit()과 같이 학습을 의미하는 것이 아니라 입력 데이터의 형태에 맞춰 데이터를 변환하기 위한 사전 구조를 맞추는 작업이다. fit()으로 변환을 위한 사전 구조를 맞추면 이후 입력 데이터의 차원 변환,클러스터링, 피처 추출 등의 실제 작업은 **transform()**으로 수행한다.
*   사이킷런은 fit()과 transform()을 하나로 결합한 **fit_transform()**도 함께 제공한다. fit_transform()은 fit()과 transform()을 별도로 호출할 필요를 줄여주지만 사용에 약간의 주의가 필요하다.

### 사이킷런의 주요 모듈
*   [예제 데이터] sklearn.datasets : 사이킷런에 내장되어 예제로 제공하는 데이터 세트
*   [피처 처리] i) sklearn.preprocessing : 데이터 전처리에 필요한 다양한 가공 기능 제공(e.g. 문자열을 숫자형 코드 값으로 인코딩, 정규화, 스케일링 등) ii) sklearn.feature_selection : 알고리즘에 큰 영향을 미치는 피처를 우선순위대로 셀렉션 작업을 수행하는 다양한 기능 제공 iii) sklearn.feature_extraction : 텍스트 데이터나 이미지 데이터의 벡터화된 피처를 추출하는 데 사용
*   [피처 처리&차원 축소] sklearn.decomposition : 차원 축소에 관련한 알고리즘을 지원하는 모듈
*   [데이터 분리, 검증 & 파라미터 튜닝] sklearn.model_selection : 교차 검증을 위한 학습용/테스트용 분리, 그리드 서치(Grid Search)로 최적 파라미터 추출 등의 API 제공
*   [평가] sklearn.metrics : 분류, 회귀, 클러스터링, 페어와이즈(Pairwise)에 대한 다양한 성능 측정 방법 제공
*   [ML 알고리즘] i) sklearn.ensemble : 앙상블 알고리즘 및 랜덤 포레스트, 에이다 부스트, 그래디언트 부스팅 등을 제공 ii) sklearn.linear_model : 주로 선형 회귀, 릿지(Ridge), 라쏘(Lasso) 및 로지스틱 회귀 등 회귀 관련 알고리즘을 지원, 또 SGD(Stochastic Gradient Descent) 관련 알고리즘도 제공 iii) sklearn.naive_bayes : 나이브 베이즈 알고리즘 제공 iv) sklearn.neighbors : 최근접 이웃 알고리즘 제공 v) sklearn.svm : 서포트 벡터 머신 알고리즘 제공 vi) sklearn.tree : 의사 결정 트리 알고리즘 제공 vii) sklearn.cluster : 비지도 클러스터링 알고리즘 제공
*   [유틸리티] sklearn.pipeline : 피처 처리 등의 변환과 ML 알고리즘 학습, 예측 등을 함께 묶어서 실행할 수 있는 유틸리티 제공
*   일반적으로 **머신러닝 모델을 구축하는 주요 프로세스**는 피처의 가공, 변경, 추출을 수행하는 피처 처리(feature processing), ML 알고리즘 학습/예측 수행, 그리고 모델 평가의 단계를 반복적으로 수행하는 것이다.


### 내장된 예제 데이터 세트

*   사이킷런에는 별도의 외부 웹사이트에서 데이터 세트를 내려받을 필요 없이 예제로 활용할 수 있는 간단하면서도 좋은 데이터 세트가 내장되어 있다. 이 데이터는 datasets 모듈에 있는 여러 API를 호출해 만들 수 있다. 사이킷런에 내장되어 있는 데이터 세트는 분류나 회귀를 연습하기 위한 예제 용도의 데이터 세트와, 분류나 클러스터링을 위해 표본 데이터로 생성될 수 있는 데이터 세트로 나뉜다.

*   사이킷런에 내장된 이 데이터 세트들은 일반적으로 딕셔너리 형태로 되어있다. 키는 보통 data, target, target_name, feature_names, DESCR로 구성되어 있다. **data**는 피처의 데이터 세트를 가리키며, **target**은 분류 시 레이블 값, 회귀일 때는 숫자 결괏값 데이터 세트이다. **target_names**는 개별 레이블의 이름을 나타내고, **feature_names**는 피처의 이름을 나타낸다. 마지막으로 **DESCR**은 데이터 세트에 대한 설명과 각 피처의 설명을 나타낸다. data, target은 넘파이 배열(ndarray) 타입이며, target_names, feature_names는 넘파이 배열 또는 파이썬 리스트 타입이다. DESCR은 스트링 타입이다. 피처의 데이터 값을 반환받기 위해서는 내장 데이터 세트 API를 호출한 뒤, 그 Key값을 지정하면 된다.






In [10]:
from sklearn.datasets import load_iris

iris_data = load_iris()
print(type(iris_data))

<class 'sklearn.utils._bunch.Bunch'>




*   Bunch 클래스는 파이썬 딕셔너리 자료형과 유사하다. 데이터 세트에 내장되어 있는 대부분의 데이터 세트는 이와 같이 딕셔너리 형태의 값을 반환한다.
*   데이터 키는 피처들의 데이터 값을 가리킨다. 데이터 세트가 딕셔너리 형태이기 때문에 피처 데이터 값을 추출하기 위해서는 데이터 세트.data(또는 데이터 세트['data'])를 이용하면 된다.



In [11]:
keys = iris_data.keys()
print('붓꽃 데이터 세트의 키들:', keys)

붓꽃 데이터 세트의 키들: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


In [12]:
print('\n feature_names의 type:', type(iris_data.feature_names))  ## feature_names : sepal length, sepal width, petal length, petal width
print(' feature_names의 shape:', len(iris_data.feature_names))
print(iris_data.feature_names)

print('\n target_names의 type:', type(iris_data.target_names))  ## target_names : setosa, versicolor, virginica
print(' target_names의 shape:', len(iris_data.target_names))
print(iris_data.target_names)

print('\n data의 type:', type(iris_data.data))
print(' data의 shape:', iris_data.data.shape)
print(iris_data['data'])

print('\n target의 type:', type(iris_data.target))
print(' target의 shape:', iris_data.target.shape)
print(iris_data.target)

## load_iris()가 반환하는 객체의 키인 feature_names, target_name, data, target이 가리키는 값을 출력한다


 feature_names의 type: <class 'list'>
 feature_names의 shape: 4
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

 target_names의 type: <class 'numpy.ndarray'>
 target_names의 shape: 3
['setosa' 'versicolor' 'virginica']

 data의 type: <class 'numpy.ndarray'>
 data의 shape: (150, 4)
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5

## Model Selection 모듈 소개



*   사이킷런의 **model_selection 모듈**은 학습 데이터와 테스트 데이터 세트를 분리하거나 교차 검증 분할 및 평가, 그리고 Estimator의 하이퍼 파라미터를 튜닝하기 위한 다양한 함수와 클래스를 제공한다.

### 학습/테스트 데이터 세트 분리 - train_test_split()
*   train_test_split()은 전체 데이터를 학습 데이터와 테스트 데이터 세트로 분리해준다.



In [13]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

## 먼저 테스트 데이터 세트를 이용하지 않고 학습 데이터 세트로만 학습하고 예측하면 무엇이 문제인지 살펴본다
iris = load_iris()
dt_clf = DecisionTreeClassifier()
train_data = iris.data
train_label = iris.target
dt_clf.fit(train_data, train_label)

## 학습 데이터 세트로 예측을 수행한다. 즉, 학습과 예측을 동일한 데이터로 수행한다
pred = dt_clf.predict(train_data)
print('예측 정확도:', accuracy_score(train_label, pred))

예측 정확도: 1.0


위의 예측 결과가 100% 정확한 이유는 이미 학습한 학습 데이터 세트를 기반으로 예측했기 때문이다. 따라서 예측을 수행하는 데이터 세트는 학습을 수행한 학습용 데이터 세트가 아닌 테스트 전용 데이터 세트여야만 한다.


*   train_test_split()는 첫 번째 파라미터로 피처 데이터 세트, 두 번째 파라미터로 레이블 데이터 세트를 입력받는다. 그리곤 선택적으로 다음 파라미터를 입력받는다. i) **test_size** : 전체 데이터에서 테스트 데이터 세트 크기를 얼마로 샘플링할 것인가를 결정하는 기능. 디폴트는 0.25 (=25%) ii) **train_size** : 전체 데이터에서 학습용 데이터 세트 크기를 얼마로 샘플링할 것인가를 결정하는 기능. test_size parameter를 통상적으로 사용하므로 train_size는 잘 사용되지 않는다 iii) **shuffle** : 데이터를 분리하기 전에 데이터를 미리 섞어 이를 분산시키고 효율적인 학습 및 테스트 데이터를 만들지 결정하는 기능. 디폴트는 True iv) **random_state** : 호출할 때마다 동일한 학습/테스트용 데이터 세트를 생성하기 위해 주어지는 난수 값. 이를 지정하지 않으면 train_test_split()는 호출 시 무작위로 데이터를 분리하므로 수행할 때마다 다른 학습/테스트 용 데이터를 생성한다.



In [14]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

## 붓꽃 데이터 세트를 train_test_split()을 이용해 테스트 데이터 세트를 전체의 30%로, 학습 데이터 세트를 70%로 분리한다. random_state=121로 변경한다

dt_clf = DecisionTreeClassifier()
iris_data = load_iris()

X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target,
                                                    test_size=0.3, random_state=121)

In [15]:
dt_clf.fit(X_train, y_train)
pred = dt_clf.predict(X_test)
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))

예측 정확도: 0.9556


붓꽃 데이터는 150개의 데이터로 데이터 양이 크지 않아, 전체의 30% 정도인 테스트 데이터는 45개 정도밖에 되지 않으므로 이를 통해 알고리즘의 예측 성능을 판단하기에는 적절치 않다.



*   학습을 위한 데이터의 양을 일정 수준 이상으로 보장하는 것도 중요하지만, 학습된 모델에 대해 다양한 데이터를 기반으로 예측 성능을 평가해보는 것도 매우 중요하다.

### 교차 검증
*   앞에서 알고리즘을 학습시키는 학습 데이터와 이에 대한 예측 성능을 평가하기 위한 별도의 테스트용 데이터가 필요하다고 말했다. 그러나 이 방법 역시 과적합(Overfitting)에 취약할 수 있다. **과적합**이란 모델이 학습 데이터에만 과도하게 최적화되어 실제 예측을 다른 데이터로 수행할 경우에는 예측 성능이 매우 떨어지는 것을 의미한다. 그런데 고정된 학습 데이터와 테스트 데이터로 평가를 하다 보면 테스트 데이터에만 최적의 성능을 발휘할 수 있도록 편향되게 모델을 유도하는 경향이 생기게 된다. 이러한 문제점을 개선하기 위한 방법이 바로 교차 검증이다.


*   본 고사가 테스트 데이터 세트에 대해 평가하는 거라면, 모의고사는 교차 검증에서 많은 학습과 검증 세트에서 알고리즘 학습과 평가를 수행하는 것이다. 교차 검증은 본고사를 치르기 전에 모의고사를 여러 번 보는 것이다.
*   ML은 데이터에 기반하고, 데이터는 이상치, 분포도, 다양한 속성값, 피처 중요도 등 여러 가지 ML에 영향을 미치는 요소를 가지고 있다. 특정 ML 알고리즘에서 최적으로 동작할 수 있도록 데이터를 선별해 학습한다면 실제 데이터 양식과는 다소 차이가 있을 것이고, 이는 결국 성능 저하로 이어질 것이다. **교차 검증**은 이러한 데이터 편중을 막기 위해 별도의 여러 세트로 구성된 학습 데이터 세트와 검증 데이터 세트에서 학습과 평가를 수행하는 것이다. 그리고 각 세트에서 수행한 평가 결과에 따라 하이퍼 파라미터 튜닝 등의 모델 최적화를 더욱 손쉽게 할 수 있다. 대부분의 ML 모델의 성능 평가는 i) 교차 검증 기반 1차 평가 ii) 최종적으로 테스트 데이터 세트에 적용해 평가의 프로세스이고, ML에 사용되는 데이터 세트를 세분화해 학습, 검증, 테스트 데이터 세트로 나눌 수 있다. 테스트 데이터 세트 외에 별도의 검증 데이터 세트를 둬 최종 평가 이전에 학습된 모델을 다양하게 평가하는 데 사용한다.









*   **K 폴드 교차 검증**은 가장 보편적으로 사용되는 교차 검증 기법이다. 먼저 K개의 데이터 폴드 세트를 만들어 K번만큼 각 폴드 세트에 학습과 검증 평가를 반복적으로 수행하는 방법이다. 먼저 데이터 세트를 K등분하고, 첫 번째 반복에서는 처음부터 K-1개등분을 학습 데이터 세트, 마지막 K번째 등분 하나를 검증 데이터 세트로 설정하고 학습 데이터 세트에서 학습 수행, 검증 데이터 세트에서 평가를 수행한다. 첫 번째 평가를 수행하고 나면 이제 두 번째 반복에서 다시 비슷한 학습과 평가 작업을 수행하는데, 이번에는 처음부터 K-2개등분까지, 그리고 마지막 K번째 등분을 학습 데이터 세트로, K-1번째 등분 하나를 검증 데이터로 서렁한다. 학습 데이터 세트와 검증 데이터 세트를 점진적으로 변경하면서 마지막 K번째까지 학습과 검증을 수행한다. K개의 예측 평가를 구했으면 이를 평균 내어 K 폴드 평가 결과로 반영한다.
*   사이킷런에서는 K 폴드 교차 검증 프로세스 구현을 위해 **KFold**와 **StratifiedKFold** 클래스를 제공한다. KFold 객체는 **split()**을 호출하면 학습용/검증용 데이터로 분할할 수 있는 인덱스를 반환한다. 실제로 학습용/검증용 데이터 추출은 반환된 인덱스를 기반으로 개발 코드에서 직접 수행해야 한다.



In [16]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
import numpy as np

iris = load_iris()
features = iris.data
label = iris.target
dt_clf = DecisionTreeClassifier(random_state=156)

## 먼저 KFold 클래스를 이용해 붓꽃 데이터 세트를 교차 검증하고 예측 정화도를 알아본다
## 5개의 폴드 세트로 분리하는 KFold 객체와 폴드 세트별 정확도를 담을 리스트 객체를 생성한다
kfold = KFold(n_splits=5)
cv_accuracy = []
print('붓꽃 데이터 세트 크기:', features.shape[0])

붓꽃 데이터 세트 크기: 150


In [17]:
n_iter = 0

## 이제 생성된 KFold의 객체의 split()을 호출해 저에 붓꽃 데이터를 5개의 폴드 데이터 세트로 분리한다
for train_index, test_index in kfold.split(features):
  ## kfold.split()으로 반환된 인덱스를 이용해 학습용/검증용 테스트 데이터를 추출한다
  X_train, X_test = features[train_index], features[test_index]
  y_train, y_test = label[train_index], label[test_index]
  ## 학습 및 예측을 수행한다
  dt_clf.fit(X_train, y_train)
  pred = dt_clf.predict(X_test)
  n_iter += 1
  ## 반복 시마다 정확도를 측정한다
  accuracy = np.round(accuracy_score(y_test, pred), 4)
  train_size = X_train.shape[0]
  test_size = X_test.shape[0]
  print('\n{0} 교차 검증 정확도 :{1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'.format(n_iter, accuracy, train_size, test_size))
  print('#{0} 검증 세트 인덱스: {1}'.format(n_iter, test_index))
  cv_accuracy.append(accuracy)

## 개별 iteration별 정확도를 합해 평균 정확도를 계산한다
print('\n ## 평균 검증 정확도:', np.mean(cv_accuracy))


1 교차 검증 정확도 :1.0, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#1 검증 세트 인덱스: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]

2 교차 검증 정확도 :0.9667, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#2 검증 세트 인덱스: [30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53
 54 55 56 57 58 59]

3 교차 검증 정확도 :0.8667, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#3 검증 세트 인덱스: [60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
 84 85 86 87 88 89]

4 교차 검증 정확도 :0.9333, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#4 검증 세트 인덱스: [ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119]

5 교차 검증 정확도 :0.7333, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#5 검증 세트 인덱스: [120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]

 ## 평균 검증 정확도: 0.9




*   교차 검증 시마다 검증 세트의 인덱스가 달라지는 것을 볼 수 있다. 검증 세트의 인덱스를 보면 교차 검증 시마다 split() 함수가 어떻게 인덱스를 할당하는지 알 수 있다. 첫번째 교차 검증에서는 0-29번까지, 두번째는 30-59번, .., 다섯번째는 120-149번으로 각각 30개의 검증 세트 인덱스를 생성했고 이를 기반으로 검증 세트를 추출했다.
*   **Stratified K 폴드**는 불균형한 분포도를 가진 레이블(결정 클래스) 데이터 집합을 위한 K 폴드 방식이다. 불균형한(imbalanced) 분포도를 가진 레이블 데이터 집합은 특정 레이블 값이 특이하게 많거나 매우 적어 값의 분포가 한쪽으로 치우치는 걸 말한다. K 폴드가 레이블 데이터 집합이 원본 데이터 집합의 레이블 분포를 학습 및 테스트 세트에 제대로 분배하지 못하는 경우의 문제를 해결해준다. 이를 위해 Stratified K 폴드는 원본 데이터의 레이블 분포를 먼저 고려한 뒤, 이 분포와 동일하게 학습과 검증 데이터 세트를 분배한다.


In [18]:
import pandas as pd

iris = load_iris()
iris_df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
iris_df['label'] = iris.target
iris_df['label'].value_counts()

,count
label,
0,50
1,50
2,50


붓꽃 데이터 세트를 간단하게 DataFrame으로 생성하고 레이블 값의 분포도를 확인한다. Setosa 품종, Versicolor 품종, Virginica 품종 모두가 50개이다.

In [19]:
kfold = KFold(n_splits=3)  ## 이슈가 발생하는 상황을 도출하기 위해 3개의 폴드 세트를 KFold로 생성한다
n_iter = 0
for train_index, test_index in kfold.split(iris_df) :
  n_iter += 1
  label_train = iris_df['label'].iloc[train_index]
  label_test = iris_df['label'].iloc[test_index]
  print('## 교차 검증: {0}'.format(n_iter))
  print('학습 레이블 데이터 분포: \n', label_train.value_counts())
  print('검증 레이블 데이터 분포: \n', label_test.value_counts())

## 교차 검증: 1
학습 레이블 데이터 분포: 
 label
1    50
2    50
Name: count, dtype: int64
검증 레이블 데이터 분포: 
 label
0    50
Name: count, dtype: int64
## 교차 검증: 2
학습 레이블 데이터 분포: 
 label
0    50
2    50
Name: count, dtype: int64
검증 레이블 데이터 분포: 
 label
1    50
Name: count, dtype: int64
## 교차 검증: 3
학습 레이블 데이터 분포: 
 label
0    50
1    50
Name: count, dtype: int64
검증 레이블 데이터 분포: 
 label
2    50
Name: count, dtype: int64


교차 검증 시마다 3개의 폴드 세트로 만들어지는 학습 레이블과 검증 레이블이 완전히 다른 값으로 추출되었다. 첫 번째 교차 검증에서, 학습 레이블은 0의 경우 전혀 학습하지 못하는 반면 검증 레이블은 0밖에 없으므로 학습 모델은 절대 0을 예측하지 못한다. 이런 유형으로 교차 검증 데이터 세트를 분할하면 검증 예측 정확도는 0이 될 수밖에 없다.


*   StratifiedKFold를 사용하는 방법은 KFold를 사용하는 방법과 비슷한데, 단 하나 큰 차이는 StratifiedKFold는 레이블 데이터 분포도에 따라 학습/검증 데이터를 나누기 때문에 split() 메서드에 인자로 피처 데이터 세트뿐만 아니라 레이블 데이터도 반드시 필요하는 것이다. 반면 KFold의 경우 레이블 데이터 세트는 인자로 입력하지 않아도 무방하다.
*   항목 추가



In [20]:
from sklearn.model_selection import StratifiedKFold

skf = StratifiedKFold(n_splits=3)
n_iter=0

for train_index, test_index in skf.split(iris_df, iris_df['label']):
  n_iter += 1
  label_train = iris_df['label'].iloc[train_index]
  label_test = iris_df['label'].iloc[test_index]
  print('## 교차 검증: {0}'.format(n_iter))
  print('학습 레이블 데이터 분포 \n', label_train.value_counts())
  print('검증 레이블 데이터 분포 \n', label_test.value_counts())

## 교차 검증: 1
학습 레이블 데이터 분포 
 label
2    34
0    33
1    33
Name: count, dtype: int64
검증 레이블 데이터 분포 
 label
0    17
1    17
2    16
Name: count, dtype: int64
## 교차 검증: 2
학습 레이블 데이터 분포 
 label
1    34
0    33
2    33
Name: count, dtype: int64
검증 레이블 데이터 분포 
 label
0    17
2    17
1    16
Name: count, dtype: int64
## 교차 검증: 3
학습 레이블 데이터 분포 
 label
0    34
1    33
2    33
Name: count, dtype: int64
검증 레이블 데이터 분포 
 label
1    17
2    17
0    16
Name: count, dtype: int64


전체 150개의 데이터에서 학습으로 100개, 검증으로 50개가 교차 검증 단계로 할당되었다. 학습 레이블과 검증 레이블 데이터 값의 분포도가 거의 동일하게 할당됐음을 알 수 있다. 이렇게 분할이 되어야 레이블 값을 모두 학습하고 이에 기반해 검증을 수행할 수 있다.

In [21]:
dt_clf = DecisionTreeClassifier(random_state=156)

skfold = StratifiedKFold(n_splits=3)
n_iter=0
cv_accuracy=[]

## StratifiedKFold의 split() 호출 시, 반드시 레이블 데이터 세트도 추가적으로 입력해야 한다
for train_index, test_index in skfold.split(features, label):
  ## split()으로 반환된 인덱스를 통해 학습용/검증용 테스트 데이터를 추출한다
  X_train, X_test = features[train_index], features[test_index]
  y_train, y_test = label[train_index], label[test_index]
  ## 학습 및 예측을 수행한다
  dt_clf.fit(X_train, y_train)
  pred = dt_clf.predict(X_test)
  ## 반복 시마다 정확도를 측정한다
  n_iter += 1
  accuracy = np.round(accuracy_score(y_test, pred), 4)
  train_size = X_train.shape[0]
  test_size = X_test.shape[0]
  print('\n #{0} 교차 검증 정확도 : {1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'.format(n_iter, accuracy, train_size, test_size))
  print('#{0} 검증 세트 인덱스: {1}'.format(n_iter, test_index))
  cv_accuracy.append(accuracy)

## 교차 검증별 정확도와 평균 정확도를 마지막으로 계산해준다
print('\n ## 교차 검증별 정확도: ', np.round(cv_accuracy, 4))
print('## 평균 검증 정확도:', np.round(np.mean(cv_accuracy), 4))


 #1 교차 검증 정확도 : 0.98, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#1 검증 세트 인덱스: [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115]

 #2 교차 검증 정확도 : 0.94, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#2 검증 세트 인덱스: [ 17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  67
  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82 116 117 118
 119 120 121 122 123 124 125 126 127 128 129 130 131 132]

 #3 교차 검증 정확도 : 0.98, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#3 검증 세트 인덱스: [ 34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  83  84
  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 133 134 135
 136 137 138 139 140 141 142 143 144 145 146 147 148 149]

 ## 교차 검증별 정확도:  [0.98 0.94 0.98]
## 평균 검증 정확도: 0.9667




*   Stratified K 폴드의 경우 원본 데이터의 레이블 분포도 특성을 반영한 학습 및 검증 데이터 세트를 만들 수 있으므로 왜곡된 레이블 데이터 세트에서는 반드시 Stratified K 폴드를 이용해 교차 검증해야한다.
*   일반적으로 분류(Classification)에서의 교차 검증은 K 폴드가 아니라 Stratified K 폴드로 분할되어야 한다. 회귀(Regression)에서는 Stratified K 폴드가 지원되지 않는다. 회귀의 결정값은 이산값 형태의 레이블이 아니라 연속된 숫자값이기 때문에 결정값별로 분포를 정하는 의미가 없기 때문이다.


*   사이킷런은 교차 검증을 좀 더 편리하게 수행할 수 있게 해주는 API는 **cross_val_score()**을 제공한다. KFold로 데이터를 학습하고 예측하는 코드를 보면 먼저 i) 폴드 세트를 설정하고 ii) for 루프에서 반복으로 학습 및 테스트 데이터의 인덱스를 추출한 뒤 iii) 반복적으로 학습과 예측을 수행하고 예측 성능을 반환한다. cross_val_score()는 이런 일련의 과정을 한꺼번에 수행해주는 API이다. estimator는 사이킷런의 분류 알고리즘 클래스인 Classifier 또는 회귀 알고리즘 클래스인 Regressor를 의미하고, X는 피처 데이터 세트, y는 레이블 데이터 세트, scoring은 예측 성능 평가 지표를 기술하며, cv는 교차 검증 폴드 수를 의미한다. cross_val_score() 수행 후 반환 값은 scoring 파라미터로 지정된 성능 지표 측정값을 배열 형태로 반환한다. cross_val_score()는 classifier가 입력되면 Stratified K 폴드 방식으로 레이블값의 분포에 따라 학습/테스트 세트를 분할한다.
*   cross_val_score() API의 선언 형태: cross_val_score(estimator, X, y=None, scoring=None, n_jobs=1, verbose=0, fit_params=None, pre_dispatch='2*n_jobs') 이 중 estimator, X, y, scoring, cv가 주요 파라미터이다.





In [22]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.datasets import load_iris

iris_data = load_iris()
dt_clf = DecisionTreeClassifier(random_state=156)

data = iris_data.data
label = iris_data.target

## 성능 지표는 정확도(accuracy), 교차 검증 세트는 3개로 한다
scores = cross_val_score(dt_clf, data, label, scoring='accuracy', cv=3)
print('교차 검증별 정확도:', np.round(scores, 4))
print('평균 검증 정확도:', np.round(np.mean(scores), 4))

교차 검증별 정확도: [0.98 0.94 0.98]
평균 검증 정확도: 0.9667


In [24]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

## 데이터를 로딩하고 학습 데이터와 테스트 데이터를 분리한다
iris_data = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target,
                                                    test_size=0.2, random_state=121)
dtree = DecisionTreeClassifier()

## 파라미터를 딕셔너리 형태로 설정한다
parameters = {'max_depth':[1, 2, 3], 'min_samples_split':[2, 3]}

In [25]:
import pandas as pd

## param_grid의 하이퍼 파라미터를 3개의 train, test set fold로 나누어 테스트를 수행한다고 설정한다.
## refit=True가 default이다. True이면 가장 좋은 파라미터 설정으로 재학습한다.
grid_dtree = GridSearchCV(dtree, param_grid=parameters, cv=3, refit=True)

## 붓꽃 학습 데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습/평가한다
grid_dtree.fit(X_train, y_train)

## GridSearchCV 결과를 추출해 DataFrame으로 변환한다
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score',
           'split0_test_score', 'split1_test_score', 'split2_test_score']]


,params,mean_test_score,rank_test_score,split0_test_score,split1_test_score,split2_test_score
0,"{'max_depth': 1, 'min_samples_split': 2}",0.700000,5,0.700,0.7,0.70
1,"{'max_depth': 1, 'min_samples_split': 3}",0.700000,5,0.700,0.7,0.70
2,"{'max_depth': 2, 'min_samples_split': 2}",0.958333,3,0.925,1.0,0.95
3,"{'max_depth': 2, 'min_samples_split': 3}",0.958333,3,0.925,1.0,0.95
4,"{'max_depth': 3, 'min_samples_split': 2}",0.975000,1,0.975,1.0,0.95
5,"{'max_depth': 3, 'min_samples_split': 3}",0.975000,1,0.975,1.0,0.95


In [26]:
print('GridSearchCV 최적 파라미터:', grid_dtree.best_params_)
print('GridSearchCV 최고 정확도:{0:.4f}'.format(grid_dtree.best_score_))

GridSearchCV 최적 파라미터: {'max_depth': 3, 'min_samples_split': 2}
GridSearchCV 최고 정확도:0.9750


In [27]:
# GridSearchCV의 refit으로 이미 학습된 estimator을 반환한다
estimator = grid_dtree.best_estimator_

# GridSearchCV의 best_estimator_는 이미 최적 학습이 됐으므로 별도 학습이 필요 없다.
pred = estimator.predict(X_test)
print('테스트 데이터 세트 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))

테스트 데이터 세트 정확도: 0.9667
